core/src/main/scala/cromwell/core/simpleton/WdlValueBuilder.scala

```diff
@@ -0,0 +1,151 @@
+package cromwell.core.simpleton
+
+import wdl4s.TaskOutput
+import wdl4s.types.{WdlArrayType, WdlMapType, WdlPrimitiveType, WdlType}
+import wdl4s.values.{WdlArray, WdlMap, WdlValue}
+
+import scala.language.postfixOps
+import cromwell.core.{JobOutput, JobOutputs}
+import cromwell.core.simpleton.WdlValueSimpleton._
+
+
+/** Builds arbitrary WdlValues from WdlValueSimpletons. */
+object WdlValueBuilder {
+
+  /**
+    * Looks for a WDL identifier possibly followed by a metacharacter and more stuff.
+    *
+    * Capture groups:
+    *
+    * <ol>
+    * <li>A WDL identifier</li>
+    * <li>Possibly a metacharacter and more stuff after the WDL identifier</li>
+    * </ol>
+    */
+  private val IdentifierAndPathPattern = "^([a-zA-Z][a-zA-Z0-9_]*)(.*)".r
+
+  /**
+    * Looks for an array element reference: square braces surrounding digits, possibly followed by a metacharacter and
+    * more stuff.
+    *
+    * Capture groups:
+    *
+    * <ol>
+    * <li>Square braces plus enclosed digits</li>
+    * <li>Only the enclosed digits</li>
+    * <li>Possibly a metacharacter and more stuff after the square brace enclosed digits</li>
+    * </ol>
+    */
+  private val ArrayElementPattern = raw"^(\[(\d+)\])(.*)".r
+
+  /**
+    * Looks for a map element reference: a colon followed by one or more non-metacharacters, possibly followed by a
+    * metacharacter and more stuff.  Any metacharacters in map keys have been escaped by `WdlValueSimpleton#escapeMeta`,
+    * which is taken account by the regular expression below.
+    *
+    * Capture groups:
+    *
+    * <ol>
+    * <li>A colon followed by one or more non-metacharacters, including `WdlValueSimpleton#escapeMeta` escaped metacharacters.</li>
+    * <li>Same as above except for the initial colon</li>
+    * <li>Possibly a metacharacter and more trailing stuff</li>
+    * </ol>
+    */
+  private val MapElementPattern = """
+        (?x)                             # Turn on comments and whitespace insensitivity.
+
+        ^(                               # Begin capture.  This outer capture grabs all the stuff that will
+                                         # be ignored as the simpleton parsing "descends" into the map.
+
+          :                              # colon
+            (                            # Begin nested capture.
+              (?:                        #   Begin nested nested non-capturing group.
+                \\\[                     #     An escaped open square brace;
```

Comment:
ToL: At this point, it's looking more like a lexing/parsing problem than a regex problem...